src/interfaces/ClientCommands.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * These are all of the client commands that Metals supports. Note that not support
+ * may vary based on the `InitializationSettings` the client sets.
+ *
+ *  - https://scalameta.org/metals/docs/editors/new-editor.html#metals-client-commands
+ */
+export enum ClientCommands {
```

Comment:
nitpick: `enum` are a discouraged feature of TypeScript: they are one of the few features that has a runtime behavior and the community is essentially moving away from them.

I would recommend using a literal strings union or an object instead.

Literal strings union:

```
type ClientCommands =
| "metals-echo-command"
| "metals-diagnostics-focus"
// ...
```

This has the minor downsides of not having pretty names (but it's the simplest option in my opinion)

If you go with an object, you can get pretty names:


```
const clientCommands = {
  EchoCommand: "metals-echo-command",
  FocusDiagnostics: "metals-diagnostics-focus",
  // ...
} as const
type ClientCommands = typeof clientCommands
```

`as const` is there to prevent TS from widening the value types to `string` and preserving the literal types instead.

This requires a bit more incantation for exhaustive matching, but it's not too bad if we have `fp-ts` as a dependency. Not sure it's worth the effort though.

Reply:
ah sorry, the first case only defines the types and not the values, but it's still valid

Reply:
No, actually, changed my mind: the second case is best if we want to have exhaustive command handling

Reply:
Gotcha. That makes sense. I originally had an object, then switched, and was overall just unsure. So I'll switch this back to an object 👍

Reply:
Cool, see also my comment in metals-vscode on how to handle the object exhaustively